### PR TITLE
[RISCV] Enable merging of external globals by default

### PR DESCRIPTION
**Pull request by asb**
This follows up #115495 by enabling merging of external globals by default, which had been left as a next step in order to make the previous change more incremental and so we can more easily narrow down on any identified regressions.

Enabling merging of external globals matches what Arm does (for non mach-o targets), though AArch64 doesn't as there were [some concerns](https://reviews.llvm.org/D61947) it might cause regressions in some cases.

Here are the results I have from specrate benchmarks (train workload, figures are execution time in seconds):
|                 | oldmerge mean | oldmerge stdev | newmerge mean | newmerge stdev | abs old vs new change | old vs new change as % |
| --------------- | ------------- | -------------- | ------------- | -------------- | --------------------- | ---------------------- |
| 500.perlbench_r | 159.6         | 0.4            | 159.7         | 0.6            | 0.1                   | 0.06%                  |
| 502.gcc_r       | 408.8         | 2.3            | 409.7         | 2.4            | 0.9                   | 0.22%                  |
| 505.mcf_r       | 619.4         | 3.6            | 620.6         | 4.1            | 1.2                   | 0.19%                  |
| 520.omnetpp_r   | 573.7         | 5              | 569.1         | 2.4            | \-4.6                 | \-0.80%                |
| 525.x264_r      | 183.9         | 0.2            | 184           | 0.4            | 0.1                   | 0.05%                  |
| 531.deepsjeng_r | 280.6         | 0.9            | 280           | 0.7            | \-0.6                 | \-0.21%                |
| 541.leela_r     | 284.6         | 1.1            | 284.5         | 0.9            | \-0.1                 | \-0.04%                |
| 557.xz_r        | 208.8         | 3              | 209.5         | 3.2            | 0.7                   | 0.34%                  |

This is on the Spacemit X60, with essentially rva22+v. omnetpp is the main one that sees a potentially positive change, though there was an annoyingly high standard deviation on that benchmark.

Looking at dhyrstone we see a positive codegen change in Proc0 which translates to a reduced instruction count, though not decreased runtime:

```
$ perf stat -r100 ./dry.oldmerge

 Performance counter stats for './dry.oldmerge' (100 runs):

            815.02 msec task-clock:u                     #    0.999 CPUs utilized            ( +-  0.02% )
                 3      context-switches:u               #    3.681 /sec                     ( +-  2.75% )
                 0      cpu-migrations:u                 #    0.000 /sec
                46      page-faults:u                    #   56.443 /sec
     1,304,052,614      cycles:u                         #    1.600 GHz                      ( +-  0.02% )
     1,601,705,197      instructions:u                   #    1.23  insn per cycle           ( +-  0.00% )
       100,021,180      branches:u                       #  122.728 M/sec                    ( +-  0.00% )
             3,427      branch-misses:u                  #    0.00% of all branches          ( +-  0.68% )

          0.815862 +- 0.000137 seconds time elapsed  ( +-  0.02% )

$ perf stat -r100 ./dry.newmerge

 Performance counter stats for './dry.newmerge' (100 runs):

            815.29 msec task-clock:u                     #    0.999 CPUs utilized            ( +-  0.00% )
                 0      context-switches:u               #    0.000 /sec
                 0      cpu-migrations:u                 #    0.000 /sec
                46      page-faults:u                    #   56.429 /sec
     1,304,461,619      cycles:u                         #    1.600 GHz                      ( +-  0.00% )
     1,501,837,638      instructions:u                   #    1.15  insn per cycle           ( +-  0.00% )
       100,021,173      branches:u                       #  122.697 M/sec                    ( +-  0.00% )
             2,813      branch-misses:u                  #    0.00% of all branches          ( +-  0.82% )

         0.8161623 +- 0.0000465 seconds time elapsed  ( +-  0.01% )
```

EDIT: I should elaborate on my overall view here / what feedback I'm looking for. At least for the tests I've run, the differences seem mainly in the noise and the reduction in instructions in dhrystone is meaningful enough (even if execution time isn't affected for the spacemit x60 microarch) that it seems sensible to move forwards. But I'd welcome any opposing thoughts / concerns, or indeed further input from people who've enabled globals merging downstream. I'd understand this probably included external globals merging too as enabled in this patch, but it would be helpful to confirm.

### CODE REVIEW

File: llvm/lib/Target/RISCV/RISCVTargetMachine.cpp [hunk 1]
```diff
@@ -51,12 +51,9 @@ static cl::opt<cl::boolOrDefault>
     EnableGlobalMerge("riscv-enable-global-merge", cl::Hidden,
                       cl::desc("Enable the global merge pass"));
 
-static cl::opt<bool> ForceEnableGlobalMergeExternalGlobals(
-    "riscv-force-enable-global-merge-external-globals", cl::Hidden,
-    cl::init(false),
-    cl::desc(
-        "If the global merge pass is enabled, force enable global merging of "
-        "external globals (overriding any logic that might disable it)"));
+static cl::opt<bool> DisableGlobalMergeExternalGlobals(
```
Comment (topperc) [inline llvm/lib/Target/RISCV/RISCVTargetMachine.cpp:54]: Is this option redundant with the `global-merge-on-external` option in the pass?

Comment (asb) [reply llvm/lib/Target/RISCV/RISCVTargetMachine.cpp]: Thank you, it is - the naming "MergeExternalByDefault" should have made me go check. I've now deleted and updated the test that relies on this - thank you!

**Review by preames — APPROVED** [review]
LGTM.

Thank you for the performance data.  I agree the performance is pretty much in the noise, but this makes sense in concept.  Given that, I think moving forward here is the right answer.

**Review by topperc — APPROVED** [review]
LGTM

**Pull request merged**